// @bpmn-io/element-template-chooser/dist/index.js
'use strict';

var ModelUtil = require('bpmn-js/lib/util/ModelUtil');

/**
 * An element template chooser that hooks into
 * properties panel fired "choose template" events.
 *
 * @param {Object} config
 * @param {EventBus} eventBus
 * @param {ElementTemplates} elementTemplates
 * @param {Translate} translate
 * @param {PopupMenu} popupMenu
 */
function ElementTemplateChooser(
    config,
    eventBus,
    elementTemplates,
    translate,
    popupMenu) {

  this._eventBus = eventBus;
  this._elementTemplates = elementTemplates;
  this._translate = translate;
  this._popupMenu = popupMenu;

  const enableChooser = !config || config.elementTemplateChooser !== false;

  enableChooser && eventBus.on('elementTemplates.select', (event) => {

    const { element } = event;

    this.open(element).then(template => {
      elementTemplates.applyTemplate(element, template);
    }).catch(err => {
      if (err !== 'user-canceled') {
        console.error('elementTemplate.select :: error', err);
      }
    });
  });
}

ElementTemplateChooser.$inject = [
  'config.connectorsExtension',
  'eventBus',
  'elementTemplates',
  'translate',
  'popupMenu'
];

ElementTemplateChooser.prototype.open = function(element) {

  const popupMenu = this._popupMenu;
  const translate = this._translate;
  const eventBus = this._eventBus;

  return new Promise((resolve, reject) => {

    const handleClosed = () => reject('user-canceled');

    eventBus.once('popupMenu.close', handleClosed);

    eventBus.once('elementTemplateChooser.chosen', event => {

      const { template } = event;

      eventBus.off('popupMenu.close', handleClosed);

      resolve(template);
    });

    popupMenu.open(element, 'element-template-chooser', { x: 0, y: 0 }, {
      title: translate('Choose element template'),
      search: true,
      width: 350
    });
  });

};

/**
 * A entry provider for the <element-template-chooser> popup menu.
 */
function ElementTemplateChooserEntryProvider(popupMenu, eventBus, translate, elementTemplates) {

  this._popupMenu = popupMenu;
  this._eventBus = eventBus;
  this._translate = translate;
  this._elementTemplates = elementTemplates;

  this.register();
}

ElementTemplateChooserEntryProvider.$inject = [
  'popupMenu',
  'eventBus',
  'translate',
  'elementTemplates'
];

/**
 * Register replace menu provider in the popup menu
 */
ElementTemplateChooserEntryProvider.prototype.register = function() {
  this._popupMenu.registerProvider('element-template-chooser', this);
};

/**
 * Adds the element templates to the replace menu.
 * @param {djs.model.Base} element
 *
 * @returns {Object}
 */
ElementTemplateChooserEntryProvider.prototype.getPopupMenuEntries = function(element) {

  // convert to object
  return this.getTemplateEntries(element).reduce((entries, [ key, value ]) => {
    entries[key] = value;

    return entries;
  }, {});

};

/**
 * Get all element templates that can be used to replace the given element.
 *
 * @param {djs.model.Base} element
 *
 * @return {Array<Object>} a list of element templates as menu entries
 */
ElementTemplateChooserEntryProvider.prototype.getTemplateEntries = function(element) {

  const eventBus = this._eventBus;
  const translate = this._translate;

  return this._getMatchingTemplates(element).map(template => {

    const {
      icon = {},
      category
    } = template;

    const entryId = `apply-template-${ template.id }`;

    return [ entryId, {
      label: template.name && translate(template.name),
      description: template.description && translate(template.description),
      documentationRef: template.documentationRef,
      imageUrl: icon.contents,
      group: category && { ...category, name: translate(category.name) },
      action: () => {
        eventBus.fire('elementTemplateChooser.chosen', { element, template });
      }
    } ];
  });
};

/**
 * Returns the templates that can the element can be replaced with.
 *
 * @param  {djs.model.Base} element
 *
 * @return {Array<ElementTemplate>}
 */
ElementTemplateChooserEntryProvider.prototype._getMatchingTemplates = function(element) {
  return this._elementTemplates.getLatest().filter(template => {
    return ModelUtil.isAny(element, template.appliesTo) && !isTemplateApplied(element, template);
  });
};


// helpers ////////////

function isTemplateApplied(element, template) {
  const businessObject = ModelUtil.getBusinessObject(element);

  if (businessObject) {
    return businessObject.get('modelerTemplate') === template.id;
  }

  return false;
}

var index = {
  __init__: [
    'elementTemplateChooser',
    'elementTemplateChooserEntryProvider'
  ],
  elementTemplateChooser: [ 'type', ElementTemplateChooser ],
  elementTemplateChooserEntryProvider: [ 'type', ElementTemplateChooserEntryProvider ]
};

module.exports = index;
