// @miragon/camunda-web-modeler/dist/editor/BpmnEditor.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import React, { useCallback, useEffect, useRef, useState } from "react";
import { Panel, PanelGroup, PanelResizeHandle } from "react-resizable-panels";
import { tss } from "tss-react";
import CustomBpmnJsModeler from "../bpmnio/bpmn/CustomBpmnJsModeler";
import { createBpmnIoEvent } from "../events/bpmnio/BpmnIoEvents";
import { createContentSavedEvent } from "../events/modeler/ContentSavedEvent";
import { createNotificationEvent } from "../events/modeler/NotificationEvent";
import { createPropertiesPanelResizedEvent } from "../events/modeler/PropertiesPanelResizedEvent";
import { createUIUpdateRequiredEvent } from "../events/modeler/UIUpdateRequiredEvent";
/**
 * The events that trigger a UI update required event.
 */
var UI_UPDATE_REQUIRED_EVENTS = [
    "import.done",
    "saveXML.done",
    "commandStack.changed",
    "selection.changed",
    "attach",
    "elements.copied",
    "propertiesPanel.focusin",
    "propertiesPanel.focusout",
    "directEditing.activate",
    "directEditing.deactivate",
    "searchPad.closed",
    "searchPad.opened",
];
/**
 * The events that trigger a content saved event.
 */
var CONTENT_SAVED_EVENT = ["import.done", "commandStack.changed"];
var useStyles = tss.create(function () { return ({
    modeler: {
        height: "100%",
    },
    propertiesPanel: {
        height: "100%",
        "&>div": {
            height: "100%",
            overflow: "auto",
        },
    },
    hidden: {
        display: "none",
    },
    modelerOnly: {
        height: "100%",
    },
}); });
var BpmnEditor = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    var _q = useStyles(), classes = _q.classes, cx = _q.cx;
    var active = props.active, xml = props.xml, onEvent = props.onEvent, className = props.className, bpmnJsOptions = props.bpmnJsOptions, modelerOptions = props.modelerOptions, propertiesPanelOptions = props.propertiesPanelOptions;
    var _r = useState(0), initializeCount = _r[0], setInitializeCount = _r[1];
    var ref = useRef(undefined);
    var handleEvent = useCallback(function (event, data) {
        // TODO: Should bpmn-js events only be forwarded if the editor is currently active?
        onEvent(createBpmnIoEvent(event, data));
        if (!active) {
            return;
        }
        if (event === "elementTemplates.errors") {
            onEvent(createNotificationEvent("Importing element templates failed. Check console for details.", "error"));
            console.error("Importing element templates failed.", data);
        }
        /**
         * If the event should trigger a UI update required event, do it.
         */
        if (event && UI_UPDATE_REQUIRED_EVENTS.includes(event)) {
            onEvent(createUIUpdateRequiredEvent(active));
        }
        /**
         * If the event should trigger a content saved event, do it.
         */
        if (event && CONTENT_SAVED_EVENT.includes(event) && ref.current) {
            ref.current
                .save()
                .then(function (saved) {
                onEvent(createContentSavedEvent(saved.xml, saved.svg, "diagram.changed"));
            })
                .catch(function (e) {
                console.warn("Could not save document", e);
            });
        }
    }, [active, onEvent]);
    /**
     * Instantiates the modeler and properties panel. Only happens once on mount.
     */
    useEffect(function () {
        var _a, _b;
        var modeler = new CustomBpmnJsModeler({
            container: (_a = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.containerId) !== null && _a !== void 0 ? _a : "#bpmnview",
            propertiesPanel: (propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.hidden)
                ? undefined
                : ((_b = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.containerId) !== null && _b !== void 0 ? _b : "#bpmnprop"),
            bpmnJsOptions: bpmnJsOptions,
        });
        ref.current = modeler;
        if (modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.refs) {
            modelerOptions.refs.forEach(function (r) {
                r.current = modeler;
            });
        }
        setInitializeCount(function (count) { return count + 1; });
        return function () {
            modeler.unregisterGlobalEventListener(handleEvent);
            modeler.destroy();
            ref.current = undefined;
            if (modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.refs) {
                modelerOptions.refs.forEach(function (r) {
                    r.current = undefined;
                });
            }
        };
    }, [
        handleEvent,
        bpmnJsOptions,
        propertiesPanelOptions,
        modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.containerId,
    ]);
    /**
     * Imports the specified XML. The following steps are executed:
     *
     * 1. Export the currently loaded XML.
     * 2. Check if it is different from the specified XML.
     * 3. Import the specified XML if it has changed.
     * 4. Show any errors or warnings that occurred during import.
     */
    var importXml = useCallback(function (newXml) { return __awaiter(void 0, void 0, void 0, function () {
        var currentXml, e_1, result, count;
        var _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    if (!ref.current) return [3 /*break*/, 5];
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, ((_a = ref.current) === null || _a === void 0 ? void 0 : _a.saveXML({
                            format: true,
                            preamble: false,
                        }))];
                case 2:
                    currentXml = _d.sent();
                    if (newXml === currentXml.xml) {
                        // XML has not changed
                        return [2 /*return*/];
                    }
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _d.sent();
                    return [3 /*break*/, 4];
                case 4:
                    try {
                        result = ref.current.importXML(newXml);
                        count = (_c = (_b = result.warnings) === null || _b === void 0 ? void 0 : _b.length) !== null && _c !== void 0 ? _c : 0;
                        if (count > 0) {
                            console.log("Imported with warnings", result.warnings);
                            onEvent(createNotificationEvent("Imported with ".concat(count, " warning").concat(count === 1 ? "" : "s", ". See console for details."), "warning"));
                        }
                    }
                    catch (e) {
                        console.error("Could not import XML", e);
                        onEvent(createNotificationEvent("Could not import changed XML. Is it invalid? See console for details.", "error"));
                    }
                    _d.label = 5;
                case 5: return [2 /*return*/];
            }
        });
    }); }, [onEvent]);
    /**
     * Imports the document XML whenever it changes.
     */
    useEffect(function () {
        if (initializeCount > 0) {
            importXml(xml).catch(function (e) {
                console.error("Could not import XML", e);
                onEvent(createNotificationEvent("Could not import changed XML. Is it invalid? See console for details.", "error"));
            });
        }
    }, [xml, importXml, initializeCount, onEvent]);
    useEffect(function () {
        var modeler = ref.current;
        if (initializeCount > 0 && modeler) {
            modeler.registerGlobalEventListener(handleEvent);
            return function () { return modeler.unregisterGlobalEventListener(handleEvent); };
        }
        return undefined;
    }, [initializeCount, handleEvent]);
    /**
     * Binds the current modeler instance to the keyboard when active and unbinds it when inactive.
     * Also dispatches an update UI event.
     */
    useEffect(function () {
        onEvent(createUIUpdateRequiredEvent(active));
        var cur = ref.current;
        active ? cur === null || cur === void 0 ? void 0 : cur.bindKeyboard() : cur === null || cur === void 0 ? void 0 : cur.unbindKeyboard();
    }, [active, onEvent]);
    /**
     * Imports the specified element templates whenever they change.
     */
    useEffect(function () {
        var _a, _b;
        if (!(propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.hidden)) {
            (_a = ref.current) === null || _a === void 0 ? void 0 : _a.importElementTemplates((_b = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.elementTemplates) !== null && _b !== void 0 ? _b : []);
        }
    }, [propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.hidden, propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.elementTemplates]);
    var onPropertiesPanelWidthChanged = useCallback(function (sizes) {
        onEvent(createPropertiesPanelResizedEvent(sizes[1]));
    }, [onEvent]);
    var modelerContainer = (_a = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.container) !== null && _a !== void 0 ? _a : (React.createElement("div", { id: "bpmnview", className: cx(classes.modeler, modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.className) }));
    var propertiesPanelContainer = (_b = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.container) !== null && _b !== void 0 ? _b : (React.createElement("div", { id: "bpmnprop", className: cx(classes.propertiesPanel, propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.className) }));
    if (propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.hidden) {
        return (React.createElement("div", { className: cx(classes.modelerOnly, className) }, modelerContainer));
    }
    return (React.createElement(PanelGroup, { className: className, direction: "horizontal", onLayout: onPropertiesPanelWidthChanged, style: {
            display: props.active ? "flex" : "none",
        } },
        React.createElement(Panel, { defaultSize: (_d = (_c = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.size) === null || _c === void 0 ? void 0 : _c.initial) !== null && _d !== void 0 ? _d : 75, maxSize: (_f = (_e = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.size) === null || _e === void 0 ? void 0 : _e.max) !== null && _f !== void 0 ? _f : 95, minSize: (_h = (_g = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.size) === null || _g === void 0 ? void 0 : _g.min) !== null && _h !== void 0 ? _h : 5 }, modelerContainer),
        React.createElement(PanelResizeHandle, { style: {
                cursor: "col-resize",
                width: "5px",
                backgroundColor: "rgba(0, 0, 0, 0.25)",
            } }),
        React.createElement(Panel, { defaultSize: (_k = (_j = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.size) === null || _j === void 0 ? void 0 : _j.initial) !== null && _k !== void 0 ? _k : 25, maxSize: (_m = (_l = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.size) === null || _l === void 0 ? void 0 : _l.max) !== null && _m !== void 0 ? _m : 95, minSize: (_p = (_o = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.size) === null || _o === void 0 ? void 0 : _o.min) !== null && _p !== void 0 ? _p : 5 }, propertiesPanelContainer)));
};
export default BpmnEditor;
