// @miragon/camunda-web-modeler/dist/editor/DmnEditor.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import React, { useCallback, useEffect, useRef, useState } from "react";
import { Panel, PanelGroup, PanelResizeHandle } from "react-resizable-panels";
import CustomDmnJsModeler from "../bpmnio/dmn/CustomDmnJsModeler";
import { createBpmnIoEvent } from "../events/bpmnio/BpmnIoEvents";
import { createContentSavedEvent } from "../events/modeler/ContentSavedEvent";
import { createDmnViewsChangedEvent } from "../events/modeler/DmnViewsChangedEvent";
import { createNotificationEvent } from "../events/modeler/NotificationEvent";
import { createPropertiesPanelResizedEvent } from "../events/modeler/PropertiesPanelResizedEvent";
import { createUIUpdateRequiredEvent } from "../events/modeler/UIUpdateRequiredEvent";
import { tss } from "tss-react";
/**
 * The events that trigger a UI update required event.
 */
var UI_UPDATE_REQUIRED_EVENTS = [
    "import.done",
    "saveXML.done",
    "attach",
    "dmn.views.changed",
    "views.changed",
    "view.contentChanged",
    "view.selectionChanged",
    "view.directEditingChanged",
    "propertiesPanel.focusin",
    "propertiesPanel.focusout",
];
/**
 * The events that trigger a content saved event.
 */
var CONTENT_SAVED_EVENT = [
    "import.done",
    "view.contentChanged",
    "dmn.views.changed",
    "views.changed",
    "elements.changed",
];
var useStyles = tss.create(function () { return ({
    modeler: {
        height: "100%",
    },
    propertiesPanel: {
        height: "100%",
        "&>div": {
            height: "100%",
        },
    },
    hidden: {
        display: "none",
    },
    modelerOnly: {
        height: "100%",
    },
}); });
var DmnEditor = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    var _q = useStyles(), classes = _q.classes, cx = _q.cx;
    var xml = props.xml, active = props.active, onEvent = props.onEvent, dmnJsOptions = props.dmnJsOptions, propertiesPanelOptions = props.propertiesPanelOptions, modelerOptions = props.modelerOptions, className = props.className;
    var _r = useState(undefined), activeView = _r[0], setActiveView = _r[1];
    var _s = useState(0), initializeCount = _s[0], setInitializeCount = _s[1];
    var ref = useRef(null);
    var handleEvent = useCallback(function (event, data) {
        // TODO: Should dmn-js events only be forwarded if the editor is currently active?
        onEvent(createBpmnIoEvent(event, data));
        if (!active) {
            return;
        }
        if (event === "views.changed") {
            setActiveView(data.activeView);
            onEvent(createDmnViewsChangedEvent(data.views, data.activeView));
        }
        /**
         * If the event should trigger a UI update required event, do it.
         */
        if (event && UI_UPDATE_REQUIRED_EVENTS.includes(event)) {
            onEvent(createUIUpdateRequiredEvent(active));
        }
        /**
         * If the event should trigger a content saved event, do it.
         */
        if (event && CONTENT_SAVED_EVENT.includes(event) && ref.current) {
            ref.current
                .save({ format: true })
                .then(function (saved) {
                // TODO: Save SVG (but which viewer?)
                onEvent(createContentSavedEvent(saved.xml, undefined, "diagram.changed"));
            })
                .catch(function (e) {
                console.warn("Could not save document", e);
            });
        }
    }, [active, onEvent]);
    var viewsChangedCallback = useCallback(function (event, data) {
        var _a, _b;
        void handleEvent(event.type, data);
        if ((_a = ref.current) === null || _a === void 0 ? void 0 : _a.getActiveViewer()) {
            void ((_b = ref.current) === null || _b === void 0 ? void 0 : _b.registerGlobalEventListener(handleEvent));
            return function () { var _a; return (_a = ref.current) === null || _a === void 0 ? void 0 : _a.unregisterGlobalEventListener(handleEvent); };
        }
        return undefined;
    }, [handleEvent]);
    /**
     * Instantiates the modeler and properties panel. Only happens once on mount.
     */
    useEffect(function () {
        var _a, _b;
        var modeler = new CustomDmnJsModeler({
            container: (_a = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.containerId) !== null && _a !== void 0 ? _a : "#dmnview",
            propertiesPanel: (propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.hidden)
                ? undefined
                : ((_b = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.containerId) !== null && _b !== void 0 ? _b : "#dmnprop"),
            dmnJsOptions: dmnJsOptions,
        });
        ref.current = modeler;
        if (modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.refs) {
            modelerOptions.refs.forEach(function (r) {
                r.current = modeler;
            });
        }
        setInitializeCount(function (cur) { return cur + 1; });
        return function () {
            if (modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.refs) {
                modelerOptions.refs.forEach(function (r) {
                    r.current = undefined;
                });
            }
            modeler.destroy();
        };
    }, [dmnJsOptions, modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.containerId, propertiesPanelOptions]);
    useEffect(function () {
        var modeler = ref.current;
        modeler === null || modeler === void 0 ? void 0 : modeler.on("views.changed", viewsChangedCallback);
        return function () { return modeler === null || modeler === void 0 ? void 0 : modeler.off("views.changed", viewsChangedCallback); };
    }, [viewsChangedCallback]);
    /**
     * Imports the specified XML. The following steps are executed:
     *
     * 1. Export the currently loaded XML.
     * 2. Check if it is different from the specified XML.
     * 3. Import the specified XML if it has changed.
     * 4. Show any errors or warnings that occurred during import.
     */
    var importXml = useCallback(function (newXml_1) {
        var args_1 = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args_1[_i - 1] = arguments[_i];
        }
        return __awaiter(void 0, __spreadArray([newXml_1], args_1, true), void 0, function (newXml, open) {
            var currentXml, e_1, result, count, e_2;
            var _a;
            if (open === void 0) { open = false; }
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!ref.current) return [3 /*break*/, 7];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, ((_a = ref.current) === null || _a === void 0 ? void 0 : _a.save({
                                format: true,
                            }))];
                    case 2:
                        currentXml = _b.sent();
                        if (newXml === currentXml.xml) {
                            // XML has not changed
                            return [2 /*return*/];
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _b.sent();
                        return [3 /*break*/, 4];
                    case 4:
                        _b.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, ref.current.import(newXml, open)];
                    case 5:
                        result = _b.sent();
                        count = result.warnings.length;
                        if (count > 0) {
                            console.log("Imported with warnings", result.warnings);
                            onEvent(createNotificationEvent("Imported with ".concat(count, " warning").concat(count === 1 ? "" : "s", ". See console for details."), "warning"));
                        }
                        return [3 /*break*/, 7];
                    case 6:
                        e_2 = _b.sent();
                        console.error("Could not import XML", e_2);
                        onEvent(createNotificationEvent("Could not import changed XML. Is it invalid? See console for details.", "error"));
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    }, [onEvent]);
    /**
     * Imports the document XML whenever it changes.
     */
    useEffect(function () {
        if (initializeCount > 0) {
            // Only open the view on first render
            void importXml(xml, initializeCount === 1);
        }
    }, [xml, importXml, initializeCount]);
    /**
     * Binds the current modeler instance to the keyboard when active and unbinds it when inactive.
     */
    useEffect(function () {
        onEvent(createUIUpdateRequiredEvent(active));
        var modeler = ref.current;
        if (active) {
            if (modeler === null || modeler === void 0 ? void 0 : modeler.getActiveViewer()) {
                modeler === null || modeler === void 0 ? void 0 : modeler.bindKeyboard();
                return function () { return modeler === null || modeler === void 0 ? void 0 : modeler.unbindKeyboard(); };
            }
            // TODO: Is this still required?
            // Fallback for first mount
            // TODO: Is there another way to do this? The way above does not work on first mount
            var timeout_1 = setTimeout(function () { return modeler === null || modeler === void 0 ? void 0 : modeler.bindKeyboard(); }, 1000);
            return function () {
                clearTimeout(timeout_1);
                modeler === null || modeler === void 0 ? void 0 : modeler.unbindKeyboard();
            };
        }
        return undefined;
    }, [active, activeView, onEvent]);
    var onPropertiesPanelWidthChanged = useCallback(function (sizes) {
        onEvent(createPropertiesPanelResizedEvent(sizes[1]));
    }, [onEvent]);
    var modelerContainer = (_a = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.container) !== null && _a !== void 0 ? _a : (React.createElement("div", { id: "dmnview", className: cx(classes.modeler, modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.className) }));
    var propertiesPanelContainer = (_b = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.container) !== null && _b !== void 0 ? _b : (React.createElement("div", { id: "dmnprop", className: cx(classes.propertiesPanel, propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.className) }));
    if (propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.hidden) {
        return (React.createElement("div", { className: cx(classes.modelerOnly, className) }, modelerContainer));
    }
    return (React.createElement(PanelGroup, { className: className, direction: "horizontal", onLayout: onPropertiesPanelWidthChanged, style: {
            display: props.active ? "flex" : "none",
        } },
        React.createElement(Panel, { defaultSize: (_d = (_c = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.size) === null || _c === void 0 ? void 0 : _c.initial) !== null && _d !== void 0 ? _d : 75, maxSize: (_f = (_e = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.size) === null || _e === void 0 ? void 0 : _e.max) !== null && _f !== void 0 ? _f : 95, minSize: (_h = (_g = modelerOptions === null || modelerOptions === void 0 ? void 0 : modelerOptions.size) === null || _g === void 0 ? void 0 : _g.min) !== null && _h !== void 0 ? _h : 5 }, modelerContainer),
        React.createElement(PanelResizeHandle, { style: {
                cursor: "col-resize",
                width: "5px",
                backgroundColor: "rgba(0, 0, 0, 0.25)",
            } }),
        React.createElement(Panel, { defaultSize: (_k = (_j = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.size) === null || _j === void 0 ? void 0 : _j.initial) !== null && _k !== void 0 ? _k : 25, maxSize: (_m = (_l = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.size) === null || _l === void 0 ? void 0 : _l.max) !== null && _m !== void 0 ? _m : 95, minSize: (_p = (_o = propertiesPanelOptions === null || propertiesPanelOptions === void 0 ? void 0 : propertiesPanelOptions.size) === null || _o === void 0 ? void 0 : _o.min) !== null && _p !== void 0 ? _p : 5 }, propertiesPanelContainer)));
};
export default DmnEditor;
