// @miragon/camunda-web-modeler/dist/bpmnio/dmn/CustomDmnJsModeler.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import "@bpmn-io/properties-panel/dist/assets/properties-panel.css";
import "dmn-js/dist/assets/diagram-js.css";
import "dmn-js/dist/assets/dmn-font/css/dmn-embedded.css";
import "dmn-js/dist/assets/dmn-js-decision-table-controls.css";
import "dmn-js/dist/assets/dmn-js-decision-table.css";
import "dmn-js/dist/assets/dmn-js-drd.css";
import "dmn-js/dist/assets/dmn-js-literal-expression.css";
import "dmn-js/dist/assets/dmn-js-shared.css";
import camundaModdleDescriptor from "camunda-dmn-moddle/resources/camunda.json";
import { DmnPropertiesPanelModule, DmnPropertiesProviderModule, } from "dmn-js-properties-panel";
import deepmerge from "deepmerge";
import diagramOriginModule from "diagram-js-origin";
import Modeler from "dmn-js/lib/Modeler";
import GlobalEventListenerUtil from "../GlobalEventListenerUtil";
var CustomDmnJsModeler = /** @class */ (function () {
    /**
     * Creates a new instance of the bpmn-js modeler.
     *
     * @param options The options to include
     */
    function CustomDmnJsModeler(options) {
        var mergedOptions = deepmerge.all([
            // The options passed by the user
            options.dmnJsOptions || {},
            // The library's default options
            {
                container: options.container,
                drd: {
                    additionalModules: [
                        diagramOriginModule,
                        {
                            __init__: ["globalEventListenerUtil"],
                            globalEventListenerUtil: ["type", GlobalEventListenerUtil],
                        },
                    ],
                },
                decisionTable: {
                    additionalModules: [
                        {
                            __init__: ["globalEventListenerUtil"],
                            globalEventListenerUtil: ["type", GlobalEventListenerUtil],
                        },
                    ],
                },
                literalExpression: {
                    additionalModules: [
                        {
                            __init__: ["globalEventListenerUtil"],
                            globalEventListenerUtil: ["type", GlobalEventListenerUtil],
                        },
                    ],
                },
                moddleExtensions: {
                    camunda: camundaModdleDescriptor,
                },
            },
            // The options required to display the properties panel (if desired)
            // prettier-ignore
            options.propertiesPanel
                ? {
                    drd: {
                        propertiesPanel: {
                            parent: options.propertiesPanel,
                        },
                        additionalModules: [
                            DmnPropertiesPanelModule,
                            DmnPropertiesProviderModule,
                        ],
                    },
                }
                : {},
        ]);
        this.modeler = new Modeler(mergedOptions);
    }
    /**
     * Saves the editor content as XML.
     */
    CustomDmnJsModeler.prototype.save = function (params) {
        return this.modeler.saveXML(params);
    };
    /**
     * Imports the specified XML.
     *
     * @param xml The XML to import
     * @param open Whether to open the view after importing
     */
    CustomDmnJsModeler.prototype.import = function (xml, open) {
        if (open === void 0) { open = true; }
        var ImportXMLError = /** @class */ (function (_super) {
            __extends(ImportXMLError, _super);
            function ImportXMLError(message, warnings) {
                var _this = _super.call(this, message) || this;
                _this.warnings = warnings;
                _this.name = "ImportXMLError";
                return _this;
            }
            return ImportXMLError;
        }(Error));
        try {
            return this.modeler.importXML(xml, { open: open });
        }
        catch (error) {
            if (error instanceof ImportXMLError) {
                console.error("Importing XML failed with warnings", error.warnings, error);
            }
            else {
                console.error("Importing XML failed", error);
            }
            throw error;
        }
    };
    CustomDmnJsModeler.prototype.get = function (param) {
        return this.modeler.get(param);
    };
    /**
     * Registers an event listener for bpmn-js.
     *
     * @param event The name of the event
     * @param handler The listener to register
     */
    CustomDmnJsModeler.prototype.on = function (event, handler) {
        return this.modeler.on(event, handler);
    };
    /**
     * Unregisters a previously registered listener for bpmn-js.
     *
     * @param event The name of the event
     * @param handler The previously registered listener to unregister
     */
    CustomDmnJsModeler.prototype.off = function (event, handler) {
        return this.modeler.off(event, handler);
    };
    /**
     * Returns the active viewer.
     */
    CustomDmnJsModeler.prototype.getActiveViewer = function () {
        return this.modeler.getActiveViewer();
    };
    /**
     * Returns all available views.
     */
    CustomDmnJsModeler.prototype.getViews = function () {
        return this.modeler.getViews();
    };
    /**
     * Returns the active view.
     */
    CustomDmnJsModeler.prototype.getActiveView = function () {
        return this.modeler.getActiveView();
    };
    /**
     * Opens the specified view.
     *
     * @param view The view to open
     */
    CustomDmnJsModeler.prototype.open = function (view) {
        try {
            return this.modeler.open(view);
        }
        catch (error) {
            if (error.warnings) {
                var e = error;
                console.error("Opening view failed with warnings", e.warnings, e.error);
            }
            else {
                console.error("Opening view failed", error);
            }
            throw error;
        }
    };
    /**
     * Destroys the modeler instance.
     */
    CustomDmnJsModeler.prototype.destroy = function () {
        this.modeler.destroy();
    };
    /**
     * Returns whether the command stack contains any actions that can be undone.
     * Can be used to determine if the undo button should be enabled or not.
     */
    CustomDmnJsModeler.prototype.canUndo = function () {
        var _a;
        return (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("commandStack").canUndo();
    };
    /**
     * Returns whether the command stack contains any actions that can be repeated.
     * Can be used to determine if the redo button should be enabled or not.
     */
    CustomDmnJsModeler.prototype.canRedo = function () {
        var _a;
        return (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("commandStack").canRedo();
    };
    /**
     * Returns the size of the current selection.
     */
    CustomDmnJsModeler.prototype.getSelectionSize = function () {
        var _a, _b, _c;
        return ((_c = (_b = (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("selection")) === null || _b === void 0 ? void 0 : _b.get()) === null || _c === void 0 ? void 0 : _c.length) || 0;
    };
    /**
     * Binds the keyboard to the curretn document.
     * Keyboard shortcuts will trigger actions in the editor after this has been called.
     */
    CustomDmnJsModeler.prototype.bindKeyboard = function () {
        var _a;
        (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("keyboard").bind(document);
    };
    /**
     * Unbinds the keyboard from the current document.
     * Keyboard shortcuts won't work anymore after this has been called.
     */
    CustomDmnJsModeler.prototype.unbindKeyboard = function () {
        var _a;
        (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("keyboard").unbind();
    };
    /**
     * Returns the current stack index.
     */
    CustomDmnJsModeler.prototype.getStackIndex = function () {
        var _a;
        return (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("commandStack")._stackIdx;
    };
    /**
     * Instructs the command stack to undo the last action.
     */
    CustomDmnJsModeler.prototype.undo = function () {
        var _a;
        return (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("commandStack").undo();
    };
    /**
     * Instructs the command stack to repeat the last undone action.
     */
    CustomDmnJsModeler.prototype.redo = function () {
        var _a;
        return (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("commandStack").redo();
    };
    /**
     * Registers a global event listener that will receive all bpmn-js events.
     *
     * @param listener The listener to register
     */
    CustomDmnJsModeler.prototype.registerGlobalEventListener = function (listener) {
        var _a;
        (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("globalEventListenerUtil").on(listener);
    };
    /**
     * Unregisters a previously registered global event listener.
     *
     * @param listener The listener to unregister
     */
    CustomDmnJsModeler.prototype.unregisterGlobalEventListener = function (listener) {
        var _a;
        (_a = this.modeler.getActiveViewer()) === null || _a === void 0 ? void 0 : _a.get("globalEventListenerUtil").off(listener);
    };
    return CustomDmnJsModeler;
}());
export default CustomDmnJsModeler;
