// @miragon/camunda-web-modeler/dist/DmnModeler.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import React, { useCallback, useEffect, useMemo, useRef, useState } from "react";
import { tss } from "tss-react";
import SvgIcon from "./components/SvgIcon";
import ToggleGroup from "./components/ToggleGroup";
import DmnEditor from "./editor/DmnEditor";
import XmlEditor from "./editor/XmlEditor";
import { isBpmnIoEvent } from "./events";
import { createContentSavedEvent, } from "./events/modeler/ContentSavedEvent";
var useStyles = tss.create(function () { return ({
    root: {
        height: "100%",
        overflow: "hidden",
    },
    modeToggle: {
        position: "absolute",
        left: "32px",
        bottom: "32px",
        backgroundColor: "rgba(255, 255, 255, 0.87)",
    },
    buttonTitle: {
        marginLeft: "0.5rem",
        textTransform: "none",
        maxWidth: "8rem",
        textOverflow: "ellipsis",
        overflow: "hidden",
        whiteSpace: "nowrap",
    },
    icon: {
        marginTop: "4px",
    },
}); });
var DmnModeler = function (props) {
    var _a = useStyles(), classes = _a.classes, cx = _a.cx;
    var onEvent = props.onEvent, className = props.className, xmlTabOptions = props.xmlTabOptions, modelerTabOptions = props.modelerTabOptions, xml = props.xml;
    var monacoRef = useRef(null);
    var modelerRef = useRef();
    var _b = useState([]), views = _b[0], setViews = _b[1];
    var _c = useState(undefined), activeView = _c[0], setActiveView = _c[1];
    useEffect(function () {
        if ((modelerTabOptions === null || modelerTabOptions === void 0 ? void 0 : modelerTabOptions.disabled) && !(xmlTabOptions === null || xmlTabOptions === void 0 ? void 0 : xmlTabOptions.disabled)) {
            setActiveView("xml");
        }
    }, [modelerTabOptions, xmlTabOptions]);
    var saveFile = useCallback(function (source, reason) { return __awaiter(void 0, void 0, void 0, function () {
        var _a, saved, saved;
        var _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    _a = source;
                    switch (_a) {
                        case "xml": return [3 /*break*/, 1];
                        case undefined: return [3 /*break*/, 2];
                    }
                    return [3 /*break*/, 3];
                case 1:
                    {
                        if (monacoRef.current) {
                            saved = ((_b = monacoRef.current) === null || _b === void 0 ? void 0 : _b.getValue()) || "";
                            onEvent(createContentSavedEvent(saved, undefined, reason));
                        }
                        return [3 /*break*/, 6];
                    }
                    _d.label = 2;
                case 2:
                    {
                        return [3 /*break*/, 6];
                    }
                    _d.label = 3;
                case 3:
                    if (!modelerRef.current) return [3 /*break*/, 5];
                    return [4 /*yield*/, ((_c = modelerRef.current) === null || _c === void 0 ? void 0 : _c.save({ format: true }))];
                case 4:
                    saved = _d.sent();
                    onEvent(createContentSavedEvent(saved.xml, undefined, reason));
                    _d.label = 5;
                case 5: return [3 /*break*/, 6];
                case 6: return [2 /*return*/];
            }
        });
    }); }, [onEvent]);
    var changeMode = useCallback(function (viewId) { return __awaiter(void 0, void 0, void 0, function () {
        var view;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!(viewId && activeView !== viewId)) return [3 /*break*/, 4];
                    if (!(viewId === "xml" || activeView === "xml")) return [3 /*break*/, 2];
                    return [4 /*yield*/, saveFile(activeView, "view.changed")];
                case 1:
                    _c.sent();
                    _c.label = 2;
                case 2:
                    setActiveView(viewId);
                    if (!(viewId !== "xml")) return [3 /*break*/, 4];
                    view = (_a = modelerRef.current) === null || _a === void 0 ? void 0 : _a.getViews().find(function (v) { return v.id === viewId; });
                    return [4 /*yield*/, (view && ((_b = modelerRef.current) === null || _b === void 0 ? void 0 : _b.open(view)))];
                case 3:
                    _c.sent();
                    _c.label = 4;
                case 4: return [2 /*return*/];
            }
        });
    }); }, [activeView, saveFile]);
    var localOnEvent = useCallback(function (event) {
        var _a;
        if (isBpmnIoEvent(event) && event.event === "views.changed" && event.data) {
            var data = event.data;
            setViews(data.views);
            if (!data.activeView) {
                var initialView = data.views[0];
                setActiveView(initialView.id);
                (_a = modelerRef.current) === null || _a === void 0 ? void 0 : _a.open(data.views[0]).then(function (result) {
                    if (result.warnings.length > 0) {
                        console.log("Opened initial view with warnings", result.warnings);
                    }
                }).catch(function (e) {
                    if (e.warnings) {
                        console.log("Failed to open initial view with warnings", e.warnings, e.error);
                    }
                });
            }
            else {
                setActiveView(data.activeView.id);
            }
        }
        onEvent(event);
    }, [onEvent]);
    var onXmlChanged = useCallback(function (value) {
        onEvent(createContentSavedEvent(value, undefined, "xml.changed"));
    }, [onEvent]);
    var modelerOptions = useMemo(function () {
        var _a;
        if (!(modelerTabOptions === null || modelerTabOptions === void 0 ? void 0 : modelerTabOptions.modelerOptions)) {
            return {
                refs: [modelerRef],
            };
        }
        return __assign(__assign({}, modelerTabOptions.modelerOptions), { refs: __spreadArray(__spreadArray([], ((_a = modelerTabOptions.modelerOptions.refs) !== null && _a !== void 0 ? _a : []), true), [modelerRef], false) });
    }, [modelerTabOptions]);
    var monacoOptions = useMemo(function () {
        var _a;
        if (!(xmlTabOptions === null || xmlTabOptions === void 0 ? void 0 : xmlTabOptions.monacoOptions)) {
            return {
                refs: [monacoRef],
            };
        }
        return __assign(__assign({}, xmlTabOptions.monacoOptions), { refs: __spreadArray(__spreadArray([], ((_a = xmlTabOptions.monacoOptions.refs) !== null && _a !== void 0 ? _a : []), true), [monacoRef], false) });
    }, [xmlTabOptions]);
    if (!xml) {
        return null;
    }
    return (React.createElement("div", { className: cx(classes.root, className) },
        !(modelerTabOptions === null || modelerTabOptions === void 0 ? void 0 : modelerTabOptions.disabled) && (React.createElement(DmnEditor, { xml: xml, active: activeView !== "xml", onEvent: localOnEvent, modelerOptions: modelerOptions, propertiesPanelOptions: modelerTabOptions === null || modelerTabOptions === void 0 ? void 0 : modelerTabOptions.propertiesPanelOptions, dmnJsOptions: modelerTabOptions === null || modelerTabOptions === void 0 ? void 0 : modelerTabOptions.dmnJsOptions, className: modelerTabOptions === null || modelerTabOptions === void 0 ? void 0 : modelerTabOptions.className })),
        !(xmlTabOptions === null || xmlTabOptions === void 0 ? void 0 : xmlTabOptions.disabled) && (React.createElement(XmlEditor, { xml: xml, monacoOptions: monacoOptions, active: activeView === "xml", onChanged: onXmlChanged })),
        !(modelerTabOptions === null || modelerTabOptions === void 0 ? void 0 : modelerTabOptions.disabled) && (React.createElement(ToggleGroup, { className: classes.modeToggle, options: __spreadArray(__spreadArray([], views.map(function (view) { return ({
                id: view.id,
                node: (React.createElement(React.Fragment, null,
                    React.createElement("span", { className: cx({
                            "dmn-icon-lasso-tool": view.type === "drd",
                            "dmn-icon-decision-table": view.type === "decisionTable",
                            "dmn-icon-literal-expression": view.type === "literalExpression",
                        }) }),
                    React.createElement("span", { title: view.name || "Unnamed", className: classes.buttonTitle }, view.name || "Unnamed"))),
            }); }), true), [
                {
                    id: "xml",
                    node: (React.createElement(SvgIcon, { className: classes.icon, path: "M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2\n                                        0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z" })),
                },
            ], false), onChange: changeMode, active: activeView !== null && activeView !== void 0 ? activeView : "" }))));
};
export default DmnModeler;
