// @miragon/camunda-web-modeler/dist/components/ToggleGroup.js
import React from "react";
import { tss } from "tss-react";
var useStyles = tss.create(function () { return ({
    root: {
        height: "40px",
        border: "1px solid #AAA",
        borderRadius: "4px",
        display: "flex",
        "&>*": {
            padding: "0px 16px !important",
            minWidth: "50px",
            border: "none",
            cursor: "pointer",
            transition: "all 400ms",
            color: "rgba(0, 0, 0, 0.54)",
            fill: "rgba(0, 0, 0, 0.54)",
            "&:hover": {
                backgroundColor: "rgba(0, 0, 0, 0.2)",
                color: "rgba(0, 0, 0, 0.87)",
                fill: "rgba(0, 0, 0, 0.87)",
            },
        },
        "&>:first-of-type": {
            borderTopLeftRadius: "4px",
            borderBottomLeftRadius: "4px",
        },
        "&>:last-child": {
            borderTopRightRadius: "4px",
            borderBottomRightRadius: "4px",
        },
    },
    active: {
        backgroundColor: "rgba(0, 0, 0, 0.15)",
        color: "rgba(0, 0, 0, 0.87)",
        fill: "rgba(0, 0, 0, 0.87)",
    },
}); });
var ToggleGroup = function (props) {
    var _a = useStyles(), classes = _a.classes, cx = _a.cx;
    return (React.createElement("div", { className: cx(classes.root, props.className) }, props.options.map(function (option) { return (React.createElement("button", { key: option.id, type: "button", className: cx(props.active === option.id && classes.active), onClick: function () { return void props.onChange(option.id); } }, option.node)); })));
};
export default ToggleGroup;
