// @miragon/camunda-web-modeler/dist/bpmnio/bpmn/CustomBpmnJsModeler.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import "@bpmn-io/properties-panel/dist/assets/properties-panel.css";
import "bpmn-js-element-templates/dist/assets/element-templates.css";
import "@bpmn-io/element-template-chooser/dist/element-template-chooser.css";
import "bpmn-js/dist/assets/bpmn-font/css/bpmn-embedded.css";
import "bpmn-js/dist/assets/diagram-js.css";
import camundaModdleDescriptor from "camunda-bpmn-moddle/resources/camunda.json";
import { BpmnPropertiesPanelModule, BpmnPropertiesProviderModule, } from "bpmn-js-properties-panel";
import { ElementTemplatesPropertiesProviderModule } from "bpmn-js-element-templates";
import ElementTemplateChooserModule from "@bpmn-io/element-template-chooser";
import Modeler from "bpmn-js/lib/Modeler";
import deepmerge from "deepmerge";
import GlobalEventListenerUtil from "../GlobalEventListenerUtil";
var CustomBpmnJsModeler = /** @class */ (function (_super) {
    __extends(CustomBpmnJsModeler, _super);
    /**
     * Creates a new instance of the bpmn-js modeler.
     *
     * @param options The options to include
     */
    function CustomBpmnJsModeler(options) {
        var mergedOptions = deepmerge.all([
            // The options passed by the user
            options.bpmnJsOptions || {},
            // The library's default options
            {
                container: options.container,
                additionalModules: [
                    {
                        __init__: ["globalEventListenerUtil"],
                        globalEventListenerUtil: ["type", GlobalEventListenerUtil],
                    },
                ],
                moddleExtensions: {
                    camunda: camundaModdleDescriptor,
                },
            },
            // The options required to display the properties panel (if desired)
            // prettier-ignore
            options.propertiesPanel
                ? {
                    propertiesPanel: {
                        parent: options.propertiesPanel,
                    },
                    additionalModules: [
                        BpmnPropertiesPanelModule,
                        BpmnPropertiesProviderModule,
                        ElementTemplatesPropertiesProviderModule,
                        ElementTemplateChooserModule,
                    ],
                }
                : {},
        ], {
            // Deprecated, but @dominikhorn93 said it's okay because it's gonna stay that way for
            // at least 5 years (or forever)
            clone: false,
        });
        return _super.call(this, mergedOptions) || this;
    }
    /**
     * Returns the injector.
     */
    CustomBpmnJsModeler.prototype.getInjector = function () {
        return this.get("injector");
    };
    /**
     * Saves the editor content as SVG and XML simultaneously.
     */
    CustomBpmnJsModeler.prototype.save = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, xml, svg;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, Promise.all([
                            this.saveXML({
                                format: true,
                                preamble: false,
                            }),
                            this.saveSVG(),
                        ])];
                    case 1:
                        _a = _b.sent(), xml = _a[0].xml, svg = _a[1].svg;
                        return [2 /*return*/, { xml: xml, svg: svg }];
                }
            });
        });
    };
    /**
     * Registers a global event listener that will receive all bpmn-js events.
     *
     * @param listener The listener to register
     */
    CustomBpmnJsModeler.prototype.registerGlobalEventListener = function (listener) {
        this.get("globalEventListenerUtil").on(listener);
    };
    /**
     * Unregisters a previously registered global event listener.
     *
     * @param listener The listener to unregister
     */
    CustomBpmnJsModeler.prototype.unregisterGlobalEventListener = function (listener) {
        this.get("globalEventListenerUtil").off(listener);
    };
    /**
     * Imports element templates into the editor.
     *
     * @param elementTemplates The element templates to import.
     */
    CustomBpmnJsModeler.prototype.importElementTemplates = function (elementTemplates) {
        this.get("elementTemplatesLoader").setTemplates(elementTemplates);
    };
    /**
     * Binds the keyboard to the current document.
     * Keyboard shortcuts will trigger actions in the editor after this has been called.
     */
    CustomBpmnJsModeler.prototype.bindKeyboard = function () {
        this.get("keyboard").bind(document);
    };
    /**
     * Unbinds the keyboard from the current document.
     * Keyboard shortcuts won't work anymore after this has been called.
     */
    CustomBpmnJsModeler.prototype.unbindKeyboard = function () {
        this.get("keyboard").unbind();
    };
    /**
     * Toggles the hand tool.
     */
    CustomBpmnJsModeler.prototype.toggleHandTool = function () {
        this.getInjector().get("handTool").toggle();
    };
    /**
     * Toggles the lasso tool.
     */
    CustomBpmnJsModeler.prototype.toggleLassoTool = function () {
        this.getInjector().get("lassoTool").toggle();
    };
    /**
     * Toggles the space tool.
     */
    CustomBpmnJsModeler.prototype.toggleSpaceTool = function () {
        this.getInjector().get("spaceTool").toggle();
    };
    /**
     * Toggles the global connect tool.
     */
    CustomBpmnJsModeler.prototype.toggleGlobalConnectTool = function () {
        this.getInjector().get("globalConnect").toggle();
    };
    /**
     * Toggles the search box.
     */
    CustomBpmnJsModeler.prototype.toggleFind = function () {
        this.getInjector().get("searchPad").toggle();
    };
    /**
     * Activates the edit label function.
     */
    CustomBpmnJsModeler.prototype.toggleEditLabel = function () {
        var selection = this.get("selection").get();
        if (selection.length > 0) {
            this.getInjector().get("directEditing").activate(selection[0]);
        }
    };
    /**
     * Expands the selection to all available elements.
     */
    CustomBpmnJsModeler.prototype.selectAll = function () {
        var canvas = this.getInjector().get("canvas");
        var elementRegistry = this.getInjector().get("elementRegistry");
        var selection = this.get("selection");
        // select all elements except for the invisible
        // root element
        var rootElement = canvas.getRootElement();
        var elements = elementRegistry.filter(function (element) { return element !== rootElement; });
        selection.select(elements);
    };
    /**
     * Removes the currently selected elements.
     */
    CustomBpmnJsModeler.prototype.removeSelected = function () {
        var modeling = this.get("modeling");
        var selectedElements = this.getInjector().get("selection").get();
        if (selectedElements.length === 0) {
            return;
        }
        modeling.removeElements(selectedElements.slice());
    };
    /**
     * Returns the size of the current selection.
     */
    CustomBpmnJsModeler.prototype.getSelectionSize = function () {
        return this.get("selection").get().length;
    };
    /**
     * Returns whether there is any element selected that can be copied.
     * Can be used to determine if the copy button should be enabled or not.
     */
    CustomBpmnJsModeler.prototype.canCopy = function () {
        return this.get("selection").get().length > 0;
    };
    /**
     * Returns whether there is any element in the clipboard that can be pasted.
     * Can be used to determine if the paste button should be enabled or not.
     */
    CustomBpmnJsModeler.prototype.canPaste = function () {
        return !this.get("clipboard").isEmpty();
    };
    /**
     * Returns the current stack index.
     */
    CustomBpmnJsModeler.prototype.getStackIndex = function () {
        return this.get("commandStack")._stackIdx;
    };
    /**
     * Returns whether the command stack contains any actions that can be undone.
     * Can be used to determine if the undo button should be enabled or not.
     */
    CustomBpmnJsModeler.prototype.canUndo = function () {
        return this.get("commandStack").canUndo();
    };
    /**
     * Returns whether the command stack contains any actions that can be repeated.
     * Can be used to determine if the redo button should be enabled or not.
     */
    CustomBpmnJsModeler.prototype.canRedo = function () {
        return this.get("commandStack").canRedo();
    };
    /**
     * Instructs the command stack to undo the last action.
     */
    CustomBpmnJsModeler.prototype.undo = function () {
        return this.get("commandStack").undo();
    };
    /**
     * Instructs the command stack to repeat the last undone action.
     */
    CustomBpmnJsModeler.prototype.redo = function () {
        return this.get("commandStack").redo();
    };
    /**
     * Resets the zoom level to its default value.
     */
    CustomBpmnJsModeler.prototype.resetZoom = function () {
        this.get("zoomScroll").reset();
    };
    return CustomBpmnJsModeler;
}(Modeler));
export default CustomBpmnJsModeler;
